server/src/main/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommand.java

```diff
@@ -0,0 +1,547 @@
+        final Settings settings = environment.settings();
+
+        final String indexName;
+        final int shardId;
+        final int fromNodeId;
+        final int toNodeId;
+
+        if (options.has(folderOption)) {
+            final Path path = getPath(folderOption.value(options)).getParent();
+            final Path shardParent = path.getParent();
+            final Path shardParentParent = shardParent.getParent();
+            final Path indexPath = path.resolve(ShardPath.INDEX_FOLDER_NAME);
+            if (Files.exists(indexPath) == false || Files.isDirectory(indexPath) == false) {
+                throw new ElasticsearchException("index directory [" + indexPath + "], must exist and be a directory");
+            }
+
+            final IndexMetaData indexMetaData =
+                IndexMetaData.FORMAT.loadLatestState(logger, namedXContentRegistry, shardParent);
+
+            final String shardIdFileName = path.getFileName().toString();
+            final String nodeIdFileName = shardParentParent.getParent().getFileName().toString();
+            if (Files.isDirectory(path) && shardIdFileName.chars().allMatch(Character::isDigit) // SHARD-ID path element check
+                && NodeEnvironment.INDICES_FOLDER.equals(shardParentParent.getFileName().toString()) // `indices` check
+                && nodeIdFileName.chars().allMatch(Character::isDigit) // NODE-ID check
+                && NodeEnvironment.NODES_FOLDER.equals(shardParentParent.getParent().getParent().getFileName().toString()) // `nodes` check
+            ) {
+                shardId = Integer.parseInt(shardIdFileName);
+                indexName = indexMetaData.getIndex().getName();
+                fromNodeId = Integer.parseInt(nodeIdFileName);
+                toNodeId = fromNodeId + 1;
+            } else {
+                throw new ElasticsearchException("Unable to resolve shard id. Wrong folder structure at [ " + path.toString()
+                    + " ], expected .../nodes/[NODE-ID]/indices/[INDEX-UUID]/[SHARD-ID]");
+            }
+        } else {
+            // otherwise resolve shardPath based on the index name and shard id
+            indexName = Objects.requireNonNull(indexNameOption.value(options), "Index name is required");
+            shardId = Objects.requireNonNull(shardIdOption.value(options), "Shard ID is required");
+
+            // resolve shard path in case of multi-node layout per environment
+            fromNodeId = 0;
+            toNodeId = NodeEnvironment.MAX_LOCAL_STORAGE_NODES_SETTING.get(settings);
+        }
+
+        // have to iterate over possibleLockId as NodeEnvironment; on a contrast to it - we have to fail if node is busy
+        for (int possibleLockId = fromNodeId; possibleLockId < toNodeId; possibleLockId++) {
+            try {
+                final NodeEnvironment.NodeLock nodeLock = new NodeEnvironment.NodeLock(possibleLockId, logger, environment, Files::exists);
+                try (Releasable nLock = nodeLock) {
```

Comment:
why use a different variable name instead of putting the line above directly here?

Reply:
++